library/common/http/dispatcher.cc

```diff
@@ -51,16 +52,60 @@ Dispatcher::DirectStream::DirectStream(envoy_stream_t stream_handle,
     : stream_handle_(stream_handle), underlying_stream_(underlying_stream),
       callbacks_(std::move(callbacks)) {}
 
-Dispatcher::Dispatcher(Event::Dispatcher& event_dispatcher,
-                       Upstream::ClusterManager& cluster_manager)
-    : event_dispatcher_(event_dispatcher), cluster_manager_(cluster_manager) {}
+Dispatcher::Dispatcher() {
+  event_dispatcher_.store(nullptr);
+  cluster_manager_.store(nullptr);
+}
+
+void Dispatcher::ready(Event::Dispatcher& event_dispatcher,
+                       Upstream::ClusterManager& cluster_manager) {
+  {
+    Thread::LockGuard lock(initialization_lock_);
+
+    for (const Event::PostCb& cb : initialization_queue_) {
+      event_dispatcher.post(cb);
+    }
+
+    // The HTTP dispatcher's post function relies on this atomic pointer to signal transition
+    // between the initialization phase -- where functors are posted to the initialization_queue_ --
+    // and ready state -- where functors are directly posted to the event_dispatcher_. Therefore,
+    // this pointer needs to be set at the end of this critical section, **after** the
+    // initialization_queue_ has been drained into the event_dispatcher_ to maintain ordering
+    // guarantees in the event_dispatcher_'s queue.
+    event_dispatcher_.store(&event_dispatcher);
+    cluster_manager_.store(&cluster_manager);
+  }
+}
+
+void Dispatcher::post(Event::PostCb callback) {
+  // If the event_dispatcher is present, dispatch the functor directly to it.
+  auto event_dispatcher = event_dispatcher_.load();
```

Comment:
Reading the code we've written through again, I think we could benefit down the road from encapsulating this logic in an EventQueue class. (I think we could also simplify the synchronized state handling as well by doing so.)

Reply:
EventQueue/CallbackQueue/InitQueue